pf/internal/convert/encoding.go

```diff
@@ -331,140 +239,95 @@ func (e *encoding) deriveDecoder(typeSpec *pschema.TypeSpec, t tftypes.Type) (De
-	}
-
-	switch typeSpec.Type {
-	case "boolean":
-		return newBoolDecoder(), nil
-	case "integer":
-		return newNumberDecoder(), nil
-	case "number":
-		return newNumberDecoder(), nil
-	case "string":
+	switch {
+	case t.Is(tftypes.String):
 		return newStringDecoder(), nil
-	case "array":
-		switch t := t.(type) {
+	case t.Is(tftypes.Number):
+		return newNumberDecoder(), nil
+	case t.Is(tftypes.Bool):
+		return newBoolDecoder(), nil
+	default:
+		switch tt := t.(type) {
+		case tftypes.Object:
+			mctx, err := pctx.Object()
+			if err != nil {
+				return nil, fmt.Errorf("issue deriving an object encoder: %w", err)
+			}
+			propertyDecoders, err := e.buildPropertyDecoders(mctx, tt)
+			if err != nil {
+				return nil, fmt.Errorf("issue deriving an object encoder: %w", err)
+			}
+			return newObjectDecoder(tt, propertyDecoders, mctx)
 		case tftypes.List:
-			elementDecoder, err := e.deriveDecoder(typeSpec.Items, t.ElementType)
+			elctx, err := pctx.Element()
+			if err != nil {
+				return nil, err
+			}
+			elementDecoder, err := e.deriveDecoder(elctx, tt.ElementType)
 			if err != nil {
 				return nil, err
 			}
 			return newListDecoder(elementDecoder)
+		case tftypes.Map:
+			elctx, err := pctx.Element()
+			if err != nil {
+				return nil, err
+			}
+			elementDecoder, err := e.deriveDecoder(elctx, tt.ElementType)
+			if err != nil {
+				return nil, err
+			}
+			return newMapDecoder(elementDecoder)
 		case tftypes.Set:
-			elementDecoder, err := e.deriveDecoder(typeSpec.Items, t.ElementType)
+			elctx, err := pctx.Element()
+			if err != nil {
+				return nil, err
+			}
+			elementDecoder, err := e.deriveDecoder(elctx, tt.ElementType)
 			if err != nil {
 				return nil, err
 			}
 			return newSetDecoder(elementDecoder)
+		case tftypes.Tuple:
+			return e.deriveTupleDecoder(pctx, tt)
 		default:
-			return nil, fmt.Errorf("expected a List or Set, got %s", t.String())
-		}
-	case "object":
-		// Ensure Map[string,X] type case
-		if !(typeSpec.AdditionalProperties != nil && typeSpec.Ref == "") {
-			return nil, fmt.Errorf("expected Ref or AdditionalProperties set")
-		}
-		mt, ok := t.(tftypes.Map)
-		if !ok {
-			return nil, fmt.Errorf("expected a Map, got %s", t.String())
-		}
-		elementDecoder, err := e.deriveDecoder(typeSpec.AdditionalProperties, mt.ElementType)
-		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("Cannot build a decoder type %v", t)
 		}
-		return newMapDecoder(elementDecoder)
-	default:
-		return nil, fmt.Errorf("Cannot build a decoder type %q", typeSpec.Type)
 	}
 }
 
 // A generic base function for deriving tuple encoders and decoders.
 //
 // It handles reference validation and property discovery.
-func deriveTupleBase[T any](
-	f func(*pschema.TypeSpec, tftypes.Type) (T, error),
-	tok tokens.Type,
-	typ *pschema.ComplexTypeSpec,
-	t tftypes.Tuple,
-) ([]T, error) {
+func deriveTupleBase[T any](pctx *schemaPropContext, f func(*schemaPropContext, tftypes.Type) (T, error),
```

Comment:
Do we need additional tests for tuples? I was slightly concerned here but it looks like existing tests pass.